File: 1934-evaluate-the-bracket-pairs-of-a-string/evaluate-the-bracket-pairs-of-a-string.py
class Solution:
    def evaluate(self, s: str, knowledge: List[List[str]]) -> str:
        d = {k: v for k, v in knowledge}
        res = []
        cur_key = []
        in_bracket = False
        
        for c in s:
            if c == '(':
                in_bracket = True
                cur_key = []
            elif c == ')':
                in_bracket = False
                key = "".join(cur_key)
                res.append(d.get(key, "?"))
            else:
                if in_bracket:
                    cur_key.append(c)
                else:
                    res.append(c)
                    
        return "".join(res)

File: 1934-evaluate-the-bracket-pairs-of-a-string/evaluate-the-bracket-pairs-of-a-string.java
class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {
        Map<String, String> dict = new HashMap<>();
        for (List<String> pair : knowledge) {
            dict.put(pair.get(0), pair.get(1));
        }

        StringBuilder result = new StringBuilder();
        StringBuilder currentKey = new StringBuilder();
        boolean inBracket = false;

        for (char c : s.toCharArray()) {
            if (c == '(') {
                inBracket = true;
                currentKey.setLength(0);
            } else if (c == ')') {
                inBracket = false;
                String key = currentKey.toString();
                if (dict.containsKey(key)) {
                    result.append(dict.get(key));
                } else {
                    result.append("?");
                }
            } else {
                if (inBracket) {
                    currentKey.append(c);
                } else {
                    result.append(c);
                }
            }
        }

        return result.toString();
    }
}